tevm/evm.go

```diff
@@ -136,46 +137,50 @@ func (s *gethState) CreateAccount(addr common.Address) {
 		s.Trace("CreateAccount", addr.String())
 	}
 	var empty Account
-	a := seth.Address(addr)
-	s.Accounts.SetAccount(&a, &empty)
+	s.mu.Lock()
+	s.Accounts.SetAccount((*seth.Address)(&addr), &empty)
+	s.mu.Unlock()
 }
 
 func (s *gethState) SubBalance(addr common.Address, v *big.Int) {
 	if s.Trace != nil {
 		s.Trace("SubBalance", addr.String(), v.String())
 	}
-	a := seth.Address(addr)
-	acct, _ := s.Accounts.GetAccount(&a)
+	s.mu.Lock()
+	acct, _ := s.Accounts.GetAccount((*seth.Address)(&addr))
 	bal := acct.Balance()
 	b := bal.Big()
 	b.Sub(b, v)
 	var newacct Account
 	copy(newacct[:], acct[:])
 	newacct.SetBalance(b)
-	s.Accounts.SetAccount(&a, &newacct)
+	s.Accounts.SetAccount((*seth.Address)(&addr), &newacct)
+	s.mu.Unlock()
 }
 
 func (s *gethState) AddBalance(addr common.Address, v *big.Int) {
 	if s.Trace != nil {
 		s.Trace("AddBalance", addr.String(), v.String())
 	}
-	a := seth.Address(addr)
-	acct, _ := s.Accounts.GetAccount(&a)
+	s.mu.Lock()
```

Comment:
The `vm.StateDB` methods will never be accessed concurrently by a correct program. `vm.EVM` is not goroutine-safe.

Reply:
A single `vm.EVM` will never be used by two goroutines, but the `StateDB` will if two calls are made concurrently. I suppose it will still be racy anyway, since on throw it will revert to an old snapshot which may discard changes made by another call.

I'll move the locking up the stack.